Vng/extended exceptions

crankycoder opened the pull request:
this adds support for overloading the built in MetlogClient methods like .exception() with plugins.  the primary usecase right now is replacing the simple exception() or error() methods with a richer sentry plugin to capture detailed stack traces.

Made a minor sphinx doc fix updating a reference to CLIENT_WRAPPER to the new CLIENT_HOLDER

Also added metlog_hostname and metlog_pid entries into all fields so that we can track where a log came from

crankycoder committed:
added hostname and pid as default values in the metlog fields
dictionary

crankycoder committed:
Merge branch 'master' of github.com:mozilla-services/metlog-py

crankycoder committed:
removed reference to CLIENT_WRAPPER and switched to CLIENT_HOLDER in
the configuration documentation

crankycoder committed:
force construction of the MetlogClient to require a sender and a logger

crankycoder committed:
added the ability to override MetlogClient instance  methods
with plugins

rafrombrc commented on metlog/client.py:
```diff
@@ -135,11 +137,14 @@ class MetlogClient(object):
         :param logger: Default `logger` value for all sent messages.
+                       This should be your application name and should
+                       not be modified for different instances of
+                       metlog within your the scope of your app.
```
Wording is maybe a tiny bit strong here. "This is commonly set to be the name of the current application, and isn't modified for different instances of metlog within the scope of the same application."

rafrombrc commented on metlog/client.py:
```diff
@@ -194,15 +203,17 @@ def send_message(self, msg):
         self.sender.send_message(msg)
 
-    def add_method(self, name, method):
+    def add_method(self, name, method, override):
```
I'd much rather have override default to False. No need to make people think about it in except in the very few cases where it actually matters.

rafrombrc commented on metlog/client.py:
```diff
@@ -241,6 +252,13 @@ def metlog(self, type, timestamp=None, logger=None, severity=None,
+        # data
+        fields['metlog_pid'] = self.pid
+        fields['metlog_hostname'] = self.hostname
+
```
Putting these in the `fields` dictionary is nice b/c it's a less invasive change, but I'm wondering if these shouldn't maybe be promoted to top level message attributes. Thoughts?

rafrombrc commented:
Sorry to lag on this review, slipped through my cracks. I made a few comments, but you'll notice that they're all bike-sheddy nits, great work overall, thanks! I'm leaning towards moving the `pid` and `hostname` fields up to the top level of the message definition (instead of being in the otherwise free-form `fields` data structure), but am willing to be convinced otherwise if you think it's not worth it.

crankycoder committed:
moved metlog_hostname and metlog_pid to top level fields in the metlog dictionary

changed add_method signature to default the override argument to False

cleaned up wording in docstrings for MetlogClient constructor

crankycoder commented:
i've made those changes.  i'm mostly indifferent to where the hostname/pid are stored as far as top-level of fields as they are new attributes and we don't  have any backwards compatibility to deal with.
